wc/wc.1
.Dd December 2008
.Dt WC 1

.Sh NAME

.Nm wc
.Nd word, line, and byte or character count

.Sh SYNOPSIS

.Nm wc
.Ar [-c|-m][-lw][file...]

.Sh DESCRIPTION

    The wc utility shall read one or more input files and, by default, write
the number of <newline>s, words, and bytes contained in each input file to
the standard output.

    The utility also shall write a total count for all named files, if more
than one input file is specified.

    The wc utility shall consider a word to be a non-zero-length string of
characters delimited by white space.

.Sh OPTIONS

    The wc utility shall conform to the Base Definitions volume of IEEE Std
1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -c
        Write to the standard output the number of bytes in each input file.
    -l
        Write to the standard output the number of <newline>s in each input
file.
    -m
        Write to the standard output the number of characters in each input
file.
    -w
        Write to the standard output the number of words in each input file.

    When any option is specified, wc shall report only the information
requested by the specified options.

.Sh OPERANDS

    The following operand shall be supported:

    file
        A pathname of an input file. If no file operands are specified, the
standard input shall be used.

.Sh STDIN

    The standard input shall be used only if no file operands are specified.
See the INPUT FILES section.

.Sh INPUT FILES

    The input files may be of any type.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of wc:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments and input files) and which characters are defined as
white space characters.
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error and informative
messages written to standard output.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    By default, the standard output shall contain an entry for each input
file of the form:

    "%d %d %d %s\n", <newlines>, <words>, <bytes>, <file>

    If the -m option is specified, the number of characters shall replace the
<bytes> field in this format.

    If any options are specified and the -l option is not specified, the
number of <newline>s shall not be written.

    If any options are specified and the -w option is not specified, the
number of words shall not be written.

    If any options are specified and neither -c nor -m is specified, the
number of bytes or characters shall not be written.

    If no input file operands are specified, no name shall be written and no
<blank>s preceding the pathname shall be written.

    If more than one input file operand is specified, an additional line
shall be written, of the same format as the other lines, except that the word
total (in the POSIX locale) shall be written instead of a pathname and the
total of each column shall be written as appropriate. Such an additional
line, if any, is written at the end of the output.

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        Successful completion.
    >0
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    Default.

The following sections are informative.
.Sh APPLICATION USAGE

    The -m option is not a switch, but an option at the same level as -c.
Thus, to produce the full default output with character counts instead of
bytes, the command required is:

    wc -mlw

.Sh EXAMPLES

    None.

.Sh RATIONALE

    The output file format pseudo- printf() string differs from the System V
version of wc:

    "%7d%7d%7d %s\n"

    which produces possibly ambiguous and unparsable results for very large
files, as it assumes no number shall exceed six digits.

    Some historical implementations use only <space>, <tab>, and <newline> as
word separators. The equivalent of the ISO C standard isspace() function is
more appropriate.

    The -c option stands for "character" count, even though it counts bytes.
This stems from the sometimes erroneous historical view that bytes and
characters are the same size. Due to international requirements, the -m
option (reminiscent of "multi-byte") was added to obtain actual character
counts.

    Early proposals only specified the results when input files were text
files. The current specification more closely matches historical practice.
(Bytes, words, and <newline>s are counted separately and the results are
written when an end-of-file is detected.)

    Historical implementations of the wc utility only accepted one argument
to specify the options -c, -l, and -w. Some of them also had multiple
occurrences of an option cause the corresponding count to be written multiple
times and had the order of specification of the options affect the order of
the fields on output, but did not document either of these. Because common
usage either specifies no options or only one option, and because none of
this was documented, the changes required by this volume of IEEE Std
1003.1-2001 should not break many historical applications (and do not break
any historical conforming applications).

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    cksum



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



wc.c
/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <locale.h>
#include <wchar.h>
#include <wctype.h>
#include <ctype.h>
#include <libpu.h>

static const char doc[] =
N_("wc - print the number of newlines, words, and bytes in files");

#define args_doc file_args_doc

static struct argp_option options[] = {
	{ "bytes", 'c', NULL, 0,
	  N_("Write to the standard output the number of bytes in each input file") },
	{ "lines", 'l', NULL, 0,
	  N_("Write to the standard output the number of <newline>s in each input file") },
	{ "words", 'w', NULL, 0,
	  N_("Write to the standard output the number of words in each input file") },
	{ "chars", 'm', NULL, 0,
	  N_("Write to the standard output the number of characters in each input file") },
	{ }
};

static int wc_post_walk(struct walker *w);
static int wc_actor(struct walker *w, const char *fn, int fd);
static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, args_doc, doc };

static struct walker walker;

enum {
	WC_CHAR		= (1 << 0),
	WC_LINE		= (1 << 1),
	WC_WORD		= (1 << 2),
	WC_ALL		= WC_CHAR | WC_LINE | WC_WORD,

	WCT_BYTE	= 0,
	WCT_CHAR	= 1,

	WC_BUFSZ	= 4096,
};

struct count_info {
	unsigned long	bytes;
	unsigned long	chars;
	unsigned long	words;
	unsigned long	lines;

	bool		in_word;
	bool		in_line;
};


static int outmask;
static int out_type;
static struct count_info totals;
static int total_files;

static char wc_buf[WC_BUFSZ];
static wchar_t wc_buf_mb[WC_BUFSZ];


static void count_buf_mb(struct count_info *info, const wchar_t *buf,
			 size_t len)
{
	while (len > 0) {
		wchar_t c;

		c = *buf;

		if (iswspace(c)) {
			if (info->in_word) {
				info->in_word = false;
				info->words++;
			}
			if (c == L'\n') {
				info->in_line = false;
				info->lines++;
			} else
				info->in_line = true;
		} else {
			info->in_word = true;
		}

		buf++;
		len--;
		info->chars++;
	}
}

static void count_buf(struct count_info *info, const char *buf, size_t len)
{
	while (len > 0) {
		char c;

		c = *buf;

		if (isspace(c)) {
			if (info->in_word) {
				info->in_word = false;
				info->words++;
			}
			if (c == '\n') {
				info->in_line = false;
				info->lines++;
			} else
				info->in_line = true;
		} else {
			info->in_word = true;
		}

		buf++;
		len--;
		info->chars++;
	}
}

static int count_fd(const char *fn, int fd, struct count_info *info)
{
	ssize_t rrc;
	int rc = 0;
	mbstate_t ps;
	int mb = (MB_CUR_MAX > 1);
	char *buf = wc_buf;

	memset(&ps, 0, sizeof(ps));
	memset(info, 0, sizeof(*info));

	while (1) {
		rrc = read(fd, buf, sizeof(wc_buf) - (wc_buf - buf));
		if (rrc < 0) {
			perror(fn);
			return 1;
		}
		if (rrc == 0)
			break;

		if (!mb)
			count_buf(info, wc_buf, rrc);
		else {
			unsigned int dist;
			const char *src;
			size_t n_wc;

			/* convert bytes to wide char string */
			src = wc_buf;
			n_wc = mbsrtowcs(wc_buf_mb, &src, rrc, &ps);

			/* iterate over wide char string */
			count_buf_mb(info, wc_buf_mb, n_wc);

			/* handle any leftover octets */
			dist = rrc - (src - buf);
			memmove(wc_buf, src, dist);
			buf = wc_buf + dist;
		}

		info->bytes += rrc;
	}

	if (info->in_word)
		info->words++;
	if (info->in_line)
		info->lines++;

	return rc;
}

static void print_info (const char *fn, struct count_info *info)
{
	char numstr[128], s[32];
	const char *std_in = "(standard input)";
	bool need_space;
	bool need_filename = (strcmp(fn, std_in) != 0);
	numstr[0] = 0;

	if (outmask & WC_LINE) {
		need_space = (need_filename || outmask & WC_WORD
					    || outmask & WC_CHAR
					    || outmask & WCT_BYTE);
		sprintf(s, "%lu%s", info->lines, need_space ? " " : "");
		strcat(numstr, s);
	}
	if (outmask & WC_WORD) {
		need_space = (need_filename || outmask & WC_CHAR
					    || outmask & WCT_BYTE);
		sprintf(s, "%lu%s", info->words, need_space ? " " : "");
		strcat(numstr, s);
	}
	if (outmask & WC_CHAR) {
		need_space = need_filename;
		if (out_type == WCT_BYTE)
			sprintf(s, "%lu%s", info->bytes,
						need_filename ? " " : "");
		else
			sprintf(s, "%lu%s", info->chars,
						need_filename ? " " : "");
		strcat(numstr, s);
	}

	printf("%s%s\n", numstr, need_filename ? fn : "");
}

static int wc_actor(struct walker *w, const char *fn, int fd)
{
	struct count_info info;
	int rc;

	if (!outmask)
		outmask = WC_ALL;

	rc = count_fd(fn, fd, &info);
	if (rc)
		return rc;

	totals.bytes += info.bytes;
	totals.words += info.words;
	totals.lines += info.lines;
	total_files++;

	print_info(fn, &info);

	return 0;
}

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	PU_OPT_BEGIN

	case 'c':
		outmask |= WC_CHAR;
		out_type = WCT_BYTE;
		break;
	case 'm':
		outmask |= WC_CHAR;
		out_type = WCT_CHAR;
		break;
	case 'l':
		outmask |= WC_LINE;
		break;
	case 'w':
		outmask |= WC_WORD;
		break;

	PU_OPT_ARG
	PU_OPT_DEFAULT
	PU_OPT_END
}

static int wc_post_walk(struct walker *w)
{
	if (total_files > 1)
		print_info(_("total"), &totals);

	return w->exit_status;
}

int main (int argc, char *argv[])
{
	walker.argp			= &argp;
	walker.flags			= WF_NO_FILES_STDIN;
	walker.post_walk		= wc_post_walk;
	walker.cmdline_fd		= wc_actor;
	return walk(&walker, argc, argv);
}
